robot/control/Src/motion-control/RobotController.cpp

```diff
@@ -0,0 +1,186 @@
+    //     }
+    // }
+
+    if (std::abs(linear_accel(0)) > 1.5) {
+        linear_accel(0) = 1.5 * sgn(linear_accel(0));
+    }
+
+    prev_linear_accel = linear_accel;
+    linear_accel(2, 0) = sp(2) - pv(2);
+
+    Eigen::Matrix<double, numStates, 1> robot_force = Eigen::Matrix<double, 3, 1>(ROBOT_MASS_X, ROBOT_MASS_Y, 30 * ROBOT_MASS_H * ROBOT_RADIUS).cwiseProduct(linear_accel);
+
+    Eigen::Matrix<double, numWheels, 1> wheel_force = RobotModel::get().BotToWheel * robot_force;
+
+    debugInfo.val[0] = linear_accel(0,0) * 1000;
+    debugInfo.val[1] = linear_accel(1,0) * 1000;
+    debugInfo.val[2] = linear_accel(2,0) * 1000;
+    debugInfo.val[3] = sp(0,0) * 1000;
+    debugInfo.val[4] = sp(1,0) * 1000;
+    debugInfo.val[5] = sp(2,0) * 1000;
+    debugInfo.val[6] = pv(0,0) * 1000;
+    debugInfo.val[7] = pv(1,0) * 1000;
+    debugInfo.val[8] = pv(2,0) * 1000;
+    debugInfo.val[9] = 420 * 1000;
+    
+    apply_wheel_force(wheel_force, RobotModel::get().BotToWheel * pv, outputs);
+
+    debugInfo.val[10] = outputs(0, 0) * 1000;
+    debugInfo.val[11] = outputs(1, 0) * 1000;
+    debugInfo.val[12] = outputs(2, 0) * 1000;
+    debugInfo.val[13] = outputs(3, 0) * 1000;
+}
+
+void RobotController::calculateWheel(Eigen::Matrix<double, numWheels, 1> pv,
+                                     Eigen::Matrix<double, numWheels, 1> sp,
+                                     Eigen::Matrix<double, numWheels, 1>& outputs) {
+    outputs = sp;
+    return;
+}
+
+bool RobotController::limitBodyAccel(const Eigen::Matrix<double, numStates, 1> finalTarget,
+                                     Eigen::Matrix<double, numStates, 1>& dampened) {
+    Eigen::Matrix<double, numStates, 1> delta = finalTarget - BodyPrevTarget;
+
+    double accel[numStates] = {maxForwardAccel, maxSideAccel, maxAngularAccel};
+
+    bool limited = false;
+
+    for (int i = 0; i < numStates; i++) {
+        // If we broke the accel limit
+        if (delta(i, 0) > dt*accel[i]) {
+            delta(i,0) = dt*accel[i];
+            limited = true;
+        }
+
+        if (delta(i, 0) < -dt*accel[i]) {
+            delta(i,0) = -dt*accel[i];
+            limited = true;
+        }
+    }
+
+    dampened = delta + BodyPrevTarget;
+
+    return limited;
+}
+
+bool RobotController::limitWheelAccel(const Eigen::Matrix<double, numWheels, 1> finalTarget,
+                                      Eigen::Matrix<double, numWheels, 1>& dampened) {
+    Eigen::Matrix<double, numWheels, 1> delta = finalTarget - WheelPrevTarget;
+
+    bool limited = false;
+
+    for (int i = 0; i < numWheels; i++) {
+        if (delta(i, 0) > dt*maxWheelAccel) {
+            // delta(i, 0) = dt*maxWheelAccel;
+            delta *= dt*maxWheelAccel / delta(i, 0);
+            limited = true;
+        }
+
+        if (delta(i, 0) < -dt*maxWheelAccel) {
+            delta *= -dt*maxWheelAccel / delta(i, 0);
+            limited = true;
+        }
+    }
+
+    dampened = delta + WheelPrevTarget;
```

Comment:
Honestly this could all just be a function templated over the size of the input matrix, in which case it would be the same as the above.